### PR TITLE
Service Catalog upgrade tester

### PR DESCRIPTION
**Pull request by jasiu001**
Related to [#4468](https://github.com/kyma-project/kyma/issues/4468)

To run test:
`./upgrade/upgrade-test.sh`
(do not turn on minikube)

### CODE REVIEW

File: upgrade/examiner/main.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+type ConfigFlag struct {
+	Action string
+}
+
+func main() {
+	// setup all configurations: envs, flags, stop channel
+	flg := readFlags()
+	cfg, err := readConfig()
+	fatalOnError(err, "while create config")
+	stop := internal.SetupChannel()
```
Comment (mszostok) [inline upgrade/examiner/main.go]: remove that pkg and use that one instead: https://github.com/kubernetes/apiserver/blob/master/pkg/server/signal.go#L30

File: upgrade/examiner/pkg/tests/cluster_service_broker/cleaner.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
+package cluster_service_broker
```
Comment (mszostok) [inline upgrade/examiner/pkg/tests/cluster_service_broker/cleaner.go]: do not name pkg like that, is not a good pattern:
https://blog.golang.org/package-names

instead use e.g. `clusterservicebroker` or even `clusterbroker`

File: upgrade/examiner/pkg/tests/cluster_service_broker/test_cluster_service_broker.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	klog.Info("Start creation process")
+	return c.execute()
+}
+
+// TestResources executes test for ClusterServiceBroker and clean resource after finish
+func (tb *TestBroker) TestResources(stop <-chan struct{}, namespace string) error {
+	c := newTester(tb.client, namespace)
+
+	defer func() {
+		err := tb.clean(stop, namespace)
```
Comment (mszostok) [inline upgrade/examiner/pkg/tests/cluster_service_broker/test_cluster_service_broker.go]: imo this should not be split into separate logic, this should be a part of the test directly as the e2e scenario:
```
		t.assertServiceBrokerIsReady,
		t.checkServiceClass,
		t.checkServicePlan,
		t.assertServiceInstanceIsReady,
		t.assertServiceBindingIsReady,
		t.removeServiceBinding,
		t.removeServiceInstance,
		t.unregisterServiceBroker,
```

File: upgrade/examiner/internal/client_storage.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"fmt"
+	sc "github.com/kubernetes-sigs/service-catalog/pkg/client/clientset_generated/clientset"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// ClientStorage stores all required clients in upgrade tests
+type ClientStorage struct {
+	client   *kubernetes.Clientset
```
Comment (mszostok) [inline upgrade/examiner/internal/client_storage.go]: nit: s/client/k8sClient

File: upgrade/examiner/internal/runner/task.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package runner
+
+// Task is a wrapper around the UpgradeTest which orchestrate it with its name
+type task struct {
```
Comment (mszostok) [inline upgrade/examiner/internal/runner/task.go]: not used

File: upgrade/scripts/minikube.sh [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    if [[ "$MINIKUBE_STATUS" == "Running" ]] &&
+       [[ "$KUBELET_STATUS" == "Running" ]] &&
+       [[ "$APISERVER_STATUS" == "Running" ]]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+function increaseFsInotifyMaxUserInstances() {
```
Comment (mszostok) [inline upgrade/scripts/minikube.sh]: not needed
no one is doing the `kubectl log -f`

File: upgrade/scripts/minikube.sh [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+# limitations under the License.
+
+set -o errexit
+
+CURRENT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+KUBERNETES_VERSION=1.11.5
+VM_DRIVER=hyperkit
+DISK_SIZE=20g
+MEMORY=8192
```
Comment (mszostok) [inline upgrade/scripts/minikube.sh]: so high?

File: upgrade/scripts/minikube.sh [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+
+CURRENT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+KUBERNETES_VERSION=1.11.5
+VM_DRIVER=hyperkit
+DISK_SIZE=20g
```
Comment (mszostok) [inline upgrade/scripts/minikube.sh]: needed?

File: upgrade/examiner/internal/readiness.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
+package internal
```
Comment (mszostok) [inline upgrade/examiner/internal/readiness.go]: put that and cfg into readiness pkg

File: upgrade/examiner/internal/client_storage.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
+package internal
```
Comment (mszostok) [inline upgrade/examiner/internal/client_storage.go]: put that into `clientutil` pkg

File: upgrade/examiner/pkg/tests/cluster_service_broker/cleaner.go (hunk 2 above)
Comment (mszostok) [inline upgrade/examiner/pkg/tests/cluster_service_broker/cleaner.go]: remove the folder `pkg` and move the tests under the `internal` directory

File: upgrade/upgrade-test.sh [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+echo "- Initialize Minikube"
+bash ${CURRENT_DIR}/scripts/minikube.sh
+
+echo "- Installing Tiller..."
+kubectl apply -f ${CURRENT_DIR}/assets/tiller.yaml
+
+bash ${CURRENT_DIR}/scripts/is-ready.sh kube-system name tiller
+
+echo "- Installing ServiceCatalog"
```
Comment (mszostok) [inline upgrade/upgrade-test.sh]: hardcode

File: upgrade/upgrade-test.sh [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+kubectl apply -n ${SC_NAMESPACE} -f ${CURRENT_DIR}/assets/prepare-upgrade-test-job.yaml
+
+export POD_LABEL="prepare-test-job=true"
+bash ${CURRENT_DIR}/scripts/test-pod-is-ready.sh
+
+echo "- Prepare upgrade job logs:"
+kubectl logs -n ${SC_NAMESPACE} $(kubectl get po -n ${SC_NAMESPACE} -l prepare-test-job=true -ojson | jq -r '.items | .[].metadata.name') -f
+
+echo "- Upgrade ServiceCatalog"
+helm upgrade ${SC_CHART_NAME} ${CURRENT_DIR}/charts/service-catalog-crd-0-3-1.tar.gz --namespace ${SC_NAMESPACE} --wait
```
Comment (mszostok) [inline upgrade/upgrade-test.sh]: upgrade based on current branch source-code

File: upgrade/upgrade-test.sh [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+echo "- Execute upgrade tests"
+kubectl apply -n ${SC_NAMESPACE} -f ${CURRENT_DIR}/assets/execute-upgrade-test-job.yaml
```
Comment (mszostok) [inline upgrade/upgrade-test.sh]: use also helm to do it, and remove the bash script `test-pod-is-ready.sh`

**Review by mszostok — COMMENTED** [review]
summary from f2f meeting:

- use `helm init --wait`
- install service catalog with helm --version, same for test-broker
- remove clean-up section from script 
- execute examiner via `go run`

File: upgrade/upgrade-test.sh [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+echo "- Prepare test resources"
+kubectl apply -f ${CURRENT_DIR}/assets/upgrade-job-rbac.yaml
+kubectl apply -n ${SC_NAMESPACE} -f ${CURRENT_DIR}/assets/prepare-upgrade-test-job.yaml
+
+export POD_LABEL="prepare-test-job=true"
+bash ${CURRENT_DIR}/scripts/test-pod-is-ready.sh
+
+echo "- Prepare upgrade job logs:"
+kubectl logs -n ${SC_NAMESPACE} $(kubectl get po -n ${SC_NAMESPACE} -l prepare-test-job=true -ojson | jq -r '.items | .[].metadata.name') -f
+
```
Comment (mszostok) [inline upgrade/upgrade-test.sh]: check if prepare test job not failed

File: upgrade/scripts/minikube.sh (hunk 7 above)
Comment (jasiu001) [reply upgrade/scripts/minikube.sh]: changed to 5g disk size and 6GB of memory, based on my `Activity Monitor`

File: upgrade/scripts/minikube.sh (hunk 8 above)
Comment (jasiu001) [reply upgrade/scripts/minikube.sh]: removed

File: upgrade/upgrade-test.sh (hunk 13 above)
Comment (jasiu001) [reply upgrade/upgrade-test.sh]: do you mean `go run`?

File: upgrade/upgrade-test.sh (hunk 14 above)
Comment (jasiu001) [reply upgrade/upgrade-test.sh]: not needed since test is run by `go run`

File: test/upgrade/examiner/internal/tests/broker/tester.go [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+func (t *tester) removeServiceInstance() error {
+	exist, err := t.serviceInstanceExist()
+	if err != nil {
+		return errors.Wrap(err, "failed during fetching ServiceInstance")
+	}
+	if !exist {
+		return nil
+	}
+	// remove `removeServiceInstanceFinalizer` method if BrokerTest will be fixed and
```
Comment (mszostok) [inline test/upgrade/examiner/internal/tests/broker/tester.go]: please add here info about related PR: https://github.com/kubernetes-sigs/service-catalog/pull/2656

**Review by mszostok — APPROVED** [review]
LGTM 🚀 
---
right now it's not working but it's expected behavior right? 
```
I0620 22:24:33.106220   33248 tester.go:31] Start test resources for ServiceBroker test
I0620 22:24:34.116105   33248 tester.go:55] ClusterServiceBroker "test-broker" not exist
```

It will work when https://github.com/kyma-project/kyma/issues/4459 will be resolved? If yes then please and to https://github.com/kyma-project/kyma/issues/4459 information that this automated test should be one more tested

**Comment by jasiu001** [comment]
> ## LGTM 🚀
> 
> right now it's not working but it's expected behavior right?
> 
> ```
> I0620 22:24:33.106220   33248 tester.go:31] Start test resources for ServiceBroker test
> I0620 22:24:34.116105   33248 tester.go:55] ClusterServiceBroker "test-broker" not exist
> ```
> 
> It will work when [kyma-project/kyma#4459](https://github.com/kyma-project/kyma/issues/4459) will be resolved? If yes then please and to [kyma-project/kyma#4459](https://github.com/kyma-project/kyma/issues/4459) information that this automated test should be one more tested

Yes, it is correct behavior, for now upgrade without migration tool does not work correctly.

**Pull request merged**